diffrax/_solver/langevin_srk.py

```diff
@@ -0,0 +1,320 @@
+        out = self._tay_coeffs_single(c)
+
+        def check_shape(coeff_leaf):
+            permitted_shapes = [c.shape + (3, 6), c.shape + (1, 6), c.shape + (6,)]
+            assert (
+                coeff_leaf.shape in permitted_shapes
+            ), f"leaf shape: {coeff_leaf.shape}, c shape: {c.shape}"
+
+        jtu.tree_map(check_shape, out)
+        return out
+
+    @staticmethod
+    def _eval_taylor(h, tay_coeffs: _Coeffs) -> _Coeffs:
+        # Multiplies the pre-computed Taylor coefficients by powers of h.
+        dtype = tay_coeffs.dtype
+        h_powers = jnp.power(h, jnp.arange(0, 6, dtype=h.dtype)).astype(dtype)
+        return jtu.tree_map(
+            lambda tay_leaf: jnp.tensordot(tay_leaf, h_powers, axes=1), tay_coeffs
+        )
+
+    def _recompute_coeffs(
+        self, h, gamma: LangevinX, tay_coeffs: PyTree[_Coeffs], state_h
+    ) -> _Coeffs:
+        def recompute_coeffs_leaf(c: ArrayLike, _tay_coeffs: _Coeffs):
+            # Used when the step-size h changes and coefficients need to be recomputed
+            # Depending on the size of h*gamma choose whether the Taylor expansion or
+            # direct computation is more accurate.
+            cond = h * c < self.taylor_threshold  # c is a leaf of gamma
+            if jnp.ndim(c) == 0:
+                return lax.cond(
+                    cond,
+                    lambda h_: self._eval_taylor(h_, _tay_coeffs),
+                    lambda h_: self._directly_compute_coeffs_leaf(h_, c),
+                    h,
+                )
+            else:
+                tay_out = self._eval_taylor(h, _tay_coeffs)
+                if cond.ndim < jtu.tree_leaves(tay_out)[0].ndim:
+                    cond = jnp.expand_dims(cond, axis=-1)
+
+                def select_tay_or_direct(dummy):
+                    fun = lambda _c: self._directly_compute_coeffs_leaf(h, _c)
+                    direct_out = vmap(fun)(c)
+
+                    def _choose(tay_leaf, direct_leaf):
+                        assert tay_leaf.ndim == direct_leaf.ndim == cond.ndim, (
+                            f"tay_leaf.ndim: {tay_leaf.ndim},"
+                            f" direct_leaf.ndim: {direct_leaf.ndim},"
+                            f" cond.ndim: {cond.ndim}"
+                        )
+                        return jnp.where(cond, tay_leaf, direct_leaf)
+
+                    return jtu.tree_map(_choose, tay_out, direct_out)
+
+                # If all entries of h*gamma are below threshold, only compute tay_out
+                # otherwise, compute both tay_out and direct_out and select the
+                # correct one for each dimension
+                return lax.cond(
+                    eqxi.unvmap_all(cond), lambda _: tay_out, select_tay_or_direct, None
+                )
+
+        tree_with_coeffs = jtu.tree_map(recompute_coeffs_leaf, gamma, tay_coeffs)
+        outer = jtu.tree_structure(gamma)
+        inner = self._coeffs_structure
```

Comment:
I don't think you actually need `self._coeffs_structure` here. You should be able to determine this automatically from what is already here.

Probably simplest is something like
```python
inner = sentinel = object()

def recompute_coeffs_leaf(...):
    out = ...
    nonlocal inner
    if inner is sentinel:
        inner = jtu.tree_structure(out)
    else:
        assert inner == jtu.tree_structure(out)
    return out

tree_with_coeffs = jtu.tree_map(recompute_coeffs_leaf, ...)
coeffs_with_tree = ...  # as before
```

Reply:
That's such a great trick! Thank you!